### PR TITLE
Implementation of 'Get job execution details' endpoint.

### PR DESCRIPTION
**Pull request by ypadron-in**

### CODE REVIEW

File: azkaban-web-server/src/main/java/cloudflow/services/ExecutionServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        JobExecution jobExecution = new JobExecution();
+        Long firstStartTime = node.getStartTime();
+
+        List<JobExecutionAttempt> attempts = new ArrayList<>();
+        for(Object o: node.getAttemptObjects()) {
+            Map<String, Object> attempt = (Map<String, Object>) o;
+            Integer id = (Integer) attempt.get(ExecutionAttempt.ATTEMPT_PARAM);
+            Long startTime = (Long) attempt.get(ExecutionAttempt.STARTTIME_PARAM);
+            Long endTime = (Long) attempt.get(ExecutionAttempt.ENDTIME_PARAM);
+            Status status = Status.valueOf((String) attempt.get(ExecutionAttempt.STATUS_PARAM));
```
Comment (sshardool) [inline azkaban-web-server/src/main/java/cloudflow/services/ExecutionServiceImpl.java]: May be move this processing to a private utility function ?

File: azkaban-web-server/src/main/java/cloudflow/services/ExecutionServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+            attempts.add(new JobExecutionAttempt(id, startTime, endTime, status));
+
+            if(id == 0) {
+                firstStartTime = startTime;
+            }
```
Comment (sshardool) [inline azkaban-web-server/src/main/java/cloudflow/services/ExecutionServiceImpl.java]: Why is this required ?

File: azkaban-web-server/src/main/java/cloudflow/servlets/Constants.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+  public static final String EXECUTION_ID_KEY = "execution_id";
+  public static final String START_TIME_KEY = "start_time";
+  public static final String END_TIME_KEY = "end_time";
+  public static final String STATUS_KEY = "status";
+  public static final String ATTEMPTS_KEY = "attempts";
+  public static final String ID_KEY = "id";
+
+
+
+  public static final String DEFAULT_500_ERROR_MESSAGE = "Ooops, something went terribly wrong!";
```
Comment (sshardool) [inline azkaban-web-server/src/main/java/cloudflow/servlets/Constants.java]: Maybe make the default message a little less severe sounding, something like "Unexpected/Internal error..." ? I believe the intention is to use it sparingly only when we don't have a more specific exception to return to the user?

File: azkaban-web-server/src/main/java/cloudflow/servlets/ExecutionServlet.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        final AzkabanWebServer server = (AzkabanWebServer) getApplication();
+        this.executorService = server.getExecutionService();
+    }
+
+    @Override
+    protected void handleGet(HttpServletRequest req, HttpServletResponse resp, Session session)
+        throws ServletException, IOException {
+
+        Matcher matcher = GET_JOB_EXECUTION_INFO_PATTERN.matcher(req.getRequestURI());
+        if(matcher.matches()) {
```
Comment (sshardool) [inline azkaban-web-server/src/main/java/cloudflow/servlets/ExecutionServlet.java]: With a strictly defined regex we may be limited in our ability to be specific about what part of the request `path` failed the validation. 
Recommend looking at method `UriTemplate::match()` that will directly give us the template variables on a path without having to define a explicit regex.

File: azkaban-web-server/src/main/java/azkaban/webapp/servlet/AbstractAzkabanServlet.java [hunk 5]
```diff
@@ -375,6 +375,21 @@ protected void writeJSON(final HttpServletResponse resp, final Object obj, final
     JSONUtils.toJSON(obj, resp.getOutputStream(), true);
   }
 
+  protected void sendErrorResponse (HttpServletResponse httpServletResponse, int status,
```
Comment (sshardool) [inline azkaban-web-server/src/main/java/azkaban/webapp/servlet/AbstractAzkabanServlet.java]: Makes sense to have these methods in the abstract class and they will be quite useful. Thanks.

File: azkaban-web-server/src/main/java/azkaban/webapp/AzkabanWebServerModule.java [hunk 6]
```diff
@@ -86,9 +88,16 @@ protected void configure() {
...
     bind(ExecutorManagerAdapter.class).to(resolveExecutorManagerAdaptorClassType());
+
+    addCloudFlowBindings();
+  }
+
+  private void addCloudFlowBindings() {
     bind(SpaceService.class).to(SpaceServiceImpl.class);
     bind(SpaceDao.class).to(SpaceDaoImpl.class);
     bind(SpaceSuperUserDao.class).to(SpaceSuperUserDaoImpl.class);
+
```
[the remark this answers is not in the document]
Comment (ypadron-in) [reply azkaban-web-server/src/main/java/azkaban/webapp/AzkabanWebServerModule.java]: This space is intentional, it's to separate Space related bindings and classes from those of Execution and there will be more bindings related to projects, flows and experiments later on that should also be separated by spaces to make them easier to read.

File: azkaban-web-server/src/main/java/azkaban/webapp/AzkabanWebServer.java [hunk 7]
```diff
@@ -576,14 +587,14 @@ public UserManager getUserManager() {
...
 
-  public SpaceService spaceService() {
-    return this.spaceService;
-  }
-
   public ObjectMapper objectMapper() {
     return this.objectMapper;
   }
 
+  public SpaceService spaceService() { return this.spaceService; }
```
[the remark this answers is not in the document]
Comment (ypadron-in) [reply azkaban-web-server/src/main/java/azkaban/webapp/AzkabanWebServer.java]: sure

File: azkaban-web-server/src/main/java/cloudflow/servlets/Constants.java (hunk 3 above)
Comment (ypadron-in) [reply azkaban-web-server/src/main/java/cloudflow/servlets/Constants.java]: I'll change the message.

File: azkaban-web-server/src/main/java/cloudflow/services/ExecutionServiceImpl.java (hunk 2 above)
Comment (ypadron-in) [reply azkaban-web-server/src/main/java/cloudflow/services/ExecutionServiceImpl.java]: In current code job start time and end time reflect those of the last execution attempt but this api defines job duration interval as:
start time = start time of the first attempt
end time = end time of the last attempt
This piece of code captures the start time of the first execution attempt.

File: azkaban-web-server/src/main/java/cloudflow/error/CloudFlowException.java [hunk 8]
```diff
@@ -0,0 +1,23 @@
+package cloudflow.error;
+
+import java.util.Arrays;
+import java.util.List;
+
+public class CloudFlowException extends Exception {
```
[the remark this answers is not in the document]
Comment (ypadron-in) [reply azkaban-web-server/src/main/java/cloudflow/error/CloudFlowException.java]: IMO being explicit about throwing an exception is part of the service contract, it raises awareness and forces the caller to do the appropriate handling which avoids errors at runtime that could lead to a bad user experience (users will see an error stacktrace instead of a 500 response).

File: azkaban-web-server/src/main/java/cloudflow/error/CloudFlowException.java [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+
+    // A list to avoid multiple api calls when there are various validation errors. Especially
+    // useful with endpoints that take several parameters.
+    private List<CloudFlowError> errors;
```
[the remark this answers is not in the document]
Comment (ypadron-in) [reply azkaban-web-server/src/main/java/cloudflow/error/CloudFlowException.java]: I haven't seen the use of lists of errors in current Azkaban code but I think it makes sense for CloudFlow. Also, I don't think there is a formal requirement for this. The intention of having a list of errors is basically to communicate multiple validation errors in a single call, reducing the number of api calls and leading to a better user experience and less iterations. This is especially useful when creating new executions. I agree that we should define its usage properly since it's optional and subjective. I'm open to discuss this further.

File: azkaban-web-server/src/main/java/cloudflow/error/ErrorCode.java [hunk 10]
```diff
@@ -0,0 +1,8 @@
+package cloudflow.error;
+
+
+
+public enum ErrorCode {
+    INVALID_PARAMETER,
+    UNEXPECTED_ERROR
+}
```
[the remark this answers is not in the document]
Comment (ypadron-in) [reply azkaban-web-server/src/main/java/cloudflow/error/ErrorCode.java]: These are service layer errors, not necessarily bound to the standard semantics and codes of http errors because we can define business logic errors.

File: azkaban-web-server/src/main/java/cloudflow/models/JobExecution.java [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+public class JobExecution {
+
+    private String executionId;
+    private Long startTime;
+    private Long endTime;
+    private Status status;
+    private List<JobExecutionAttempt> attempts;
+
+
+    public JobExecution () { }
```
[the remark this answers is not in the document]
Comment (ypadron-in) [reply azkaban-web-server/src/main/java/cloudflow/models/JobExecution.java]: yes

File: azkaban-web-server/src/main/java/cloudflow/services/ExecutionServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+    private static final Logger logger = LoggerFactory.getLogger(ExecutionServiceImpl.class);
+    private ExecutorManagerAdapter executorManager;
+
+    @Inject
+    public ExecutionServiceImpl(ExecutorManagerAdapter executorManager) {
+        this.executorManager = executorManager;
+    }
+
+    @Override
+    public Optional<JobExecution> getJobExecution (String executionId, String jobDefinitionId,
```
[the remark this answers is not in the document]
Comment (ypadron-in) [reply azkaban-web-server/src/main/java/cloudflow/services/ExecutionServiceImpl.java]: It depends, when we query the DB for an inexistent record we get an empty list not an exception. The approach I'm following here is that one and leaving to the clients of the service (for example: other services or a web server) the decision of declaring it an exceptional situation or not.

File: azkaban-web-server/src/main/java/cloudflow/services/ExecutionServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+        String user) throws CloudFlowException {
+        // TODO: check user permissions
```
[the remark this answers is not in the document]
Comment (ypadron-in) [reply azkaban-web-server/src/main/java/cloudflow/services/ExecutionServiceImpl.java]: Most of these will be addressed in a week or so as soon as the dependencies (projects, flows and jobs) are figured out.
I wouldn't worry too much about the rest since these implementations will likely not be compatible with the API framework of choice.

File: azkaban-web-server/src/main/java/cloudflow/services/ExecutionServiceImpl.java (hunk 1 above)
Comment (ypadron-in) [reply azkaban-web-server/src/main/java/cloudflow/services/ExecutionServiceImpl.java]: I'll do that, thanks.

File: azkaban-web-server/src/main/java/cloudflow/services/ExecutionServiceImpl.java (hunk 12 above)
Comment (li-gsalia) [inline azkaban-web-server/src/main/java/cloudflow/services/ExecutionServiceImpl.java]: We should look into breaking down this function into smaller functions later on. With all the TODOs it will become quite long.

File: azkaban-web-server/src/main/java/cloudflow/services/ExecutionServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+        }
+
+        if( executableFlow == null) {
+            logger.info("Execution with id {} wasn't found.", execId);
+            return Optional.empty();
+        }
+
+        // TODO: get job and job path given a definition id
+        String jobPath = jobDefinitionId;
+        String[] nodesInPath = jobPath.split(":");
```
[the remark this answers is not in the document]
Comment (ypadron-in) [reply azkaban-web-server/src/main/java/cloudflow/services/ExecutionServiceImpl.java]: I added examples. A path represents a single job.

File: azkaban-web-server/src/main/java/cloudflow/services/ExecutionServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+                    break;
+                }
+            }
+            if (i >= nodesToScan.size()) {
+                logger.info("Job with id {} and path {} wasn't found in execution {}.",
+                    jobDefinitionId, jobPath, execId);
+                return Optional.empty();
+            }
+            if (node instanceof ExecutableFlowBase) {
+                nodesToScan = ((ExecutableFlowBase) node).getExecutableNodes();
```
[the remark this answers is not in the document]
Comment (ypadron-in) [reply azkaban-web-server/src/main/java/cloudflow/services/ExecutionServiceImpl.java]: I'm only navigating the path to the job not doing a full traversal of the graph, so given a path `1/2/3` when I find node 1 I don't need to keep checking its siblings but rather look for node 2 in 1's children.

File: azkaban-web-server/src/main/java/cloudflow/services/ExecutionServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+            } else {
+                // If we find a job before the end of the path this ensures the next element in
+                // the path will not be found in the next iteration. This has no effect with the
+                // last node in the path.
+                nodesToScan = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (ypadron-in) [reply azkaban-web-server/src/main/java/cloudflow/services/ExecutionServiceImpl.java]: As the comment explains if provided a wrong path (for example: `embeddedFlow:job1:job2`) this could be reached before the end of a path (in job1), executing a break here would imply the node was found which is not correct.

File: azkaban-web-server/src/main/java/cloudflow/services/ExecutionServiceImpl.java (hunk 12 above)
Comment (ypadron-in) [reply azkaban-web-server/src/main/java/cloudflow/services/ExecutionServiceImpl.java]: I agree.

File: azkaban-web-server/src/main/java/azkaban/webapp/AzkabanWebServerModule.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (sshardool) [reply azkaban-web-server/src/main/java/azkaban/webapp/AzkabanWebServerModule.java]: Makes sense!

File: azkaban-web-server/src/main/java/cloudflow/services/ExecutionServiceImpl.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (sshardool) [reply azkaban-web-server/src/main/java/cloudflow/services/ExecutionServiceImpl.java]: Thanks for the explanation. Please check if the loop variable `i` needs to be reset to 0 once `nodesToScan` has been overwritten.
